### PR TITLE
Add page for choosing a web application architecture

### PR DESCRIPTION
**Pull request by echappen**
This page came out of discussions at Engineering Practices Guild meetings ([meeting notes](https://docs.google.com/document/d/15Yk92auwqTse1MaRzro_ChIFTm4Q8sCocsZIdY9dG94/edit?usp=sharing)) ([mural board](https://app.mural.co/t/gsa6/m/gsa6/1597938543224/007c0e0058e2299b244329b4f922151fb78722ef)) about using Single Page Applications for government websites.

The topic expanded into a general survey of static and server-rendered options, and the layers of interactivity that could sit on top of them. 

We hope to present this PR/page at a future Engineering Practices Guild meeting for discussion. This text could also use some Plain Language love in collaboration with content folks. Feedback is welcome by all!

This page was drafted by @alexsoble @adunkman @KendrickDaniel @echappen

### CODE REVIEW

File: _guide/_pages/web-architecture.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
+---
+title: Choosing a Web App Architecture
+sidenav: tools
+sticky_sidenav: true
+---
+
+The goal of this guide is to help you decide how to approach an application’s UI, driving towards simplicity.
```
Comment (ryanhofdotgov) [inline _guide/_pages/web-architecture.md:7]:
```suggestion
The goal of this guide is to help you decide how to approach a web application’s architecture, driving towards simplicity.
```

File: _guide/_pages/web-architecture.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+- If you can make it a static site, you should.
+- If you can’t, it should probably be a server-rendered app.
+- If your use case requires a bit of client-side interactivity, use the above options with a bit of JavaScript.
+- If your use case requires complex client-side interactivity, then you may need a single-page application.
+
+<ins>Web applications can and do shift approaches over time.</ins>
+
+Many web apps begin their life cycles with server-side rendering only, and add more client-side functionality over time in response to user behavior. Consider whether your application could initially launch as a server-rendered app, with the potential to add more client-side functionality based on user behavior.
+
+See below for examples and heuristics to help you decide which UI architecture could make the most sense for your project as a starting point:
```
Comment (ryanhofdotgov) [inline _guide/_pages/web-architecture.md:36]:
```suggestion
See below for examples and heuristics to help you decide which architecture could make the most sense for your project as a starting point:
```

File: _guide/_pages/web-architecture.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+### If you can make it a static site, you should.
+
+_When thinking about a static site, you might use words like: Jekyll, Hugo, Federalist, static HTML._
```
Comment (ryanhofdotgov) [inline _guide/_pages/web-architecture.md:40]:
```suggestion
_When thinking about a static site, you might use words like: [Jekyll](https://jekyllrb.com), [Hugo](https://gohugo.io), [Federalist](https://federalist.18f.gov), static HTML._
```

File: _guide/_pages/web-architecture.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+#### When you might need something more complex:
+
+- When your app needs authentication, user roles or permissions
+- When your app needs to draw from live data feeds or APIs
+- When your app needs to handle sensitive data or PII
+
+### If you can’t, it should probably be a server-rendered app.
+
+_When thinking about a server-rendered app, you might use words like: Django, Rails._
```
Comment (ryanhofdotgov) [inline _guide/_pages/web-architecture.md:64]:
```suggestion
_When thinking about a server-rendered app, you might use words like: [Django](https://www.djangoproject.com/), [Rails](https://rubyonrails.org/)._
```

File: _guide/_pages/web-architecture.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+#### Drawbacks to this approach:
+
+- Applications with servers and databases will need their own ATO
+- Deployment is more complex and requires more skills to maintain
+- Zero downtime deployments are more complex
+
+### If your use case requires a bit of client-side interactivity, use the above options with a bit of JavaScript.
+
+_You might use words like: Stimulus, jQuery, Plain JavaScript._
```
Comment (ryanhofdotgov) [inline _guide/_pages/web-architecture.md:86]:
```suggestion
_You might use words like: [Stimulus](https://stimulus.hotwire.dev), [jQuery](https://jquery.com), Plain JavaScript._
```

File: _guide/_pages/web-architecture.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+- Forms with complex state are harder to manage
+- Build/deployment includes both server-side build patterns and client-side build patterns
+- Hard to know where something is rendered
+- No clear conventions, no standard way to build them.
+- Can quickly turn into a ball of JavaScript with mixed frameworks
+- Generally uses two (or more) package managers
+
+### If your use case requires complex client-side interactivity, then you may need a single-page application.
+
+_You might use words like: React, React Router, Redux, Angular, Gatsby, Vue.js, Ember_
```
Comment (ryanhofdotgov) [inline _guide/_pages/web-architecture.md:108]:
```suggestion
_You might use words like: [React](https://engineering.18f.gov/javascript/frameworks/#react), [React Router](https://reactrouter.com), [Redux](https://redux.js.org), [Angular](https://engineering.18f.gov/javascript/frameworks/#angular), [Gatsby](https://www.gatsbyjs.com), [Vue.js](https://vuejs.org), [Ember](https://emberjs.com)_
```

File: _guide/_pages/web-architecture.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+### If you can’t, it should probably be a server-rendered app.
+
+_When thinking about a server-rendered app, you might use words like: Django, Rails._
+
+#### Benefits to this approach:
+
+- Stable, tried-and-true tooling
+- Only one set of development skills needed, as opposed to separate back-end and front-end development skills
+- Faster development velocity and lower costs to build and maintain than an equivalent project with separate front-end and back-end apps
+- Can use tools like Cloud.gov
```
Comment (ryanhofdotgov) [inline _guide/_pages/web-architecture.md:71]:
```suggestion
- Can use tools like [Cloud.gov](https://cloud.gov)
```

File: _guide/_pages/web-architecture.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+#### Drawbacks to this approach:
+
+- Forms with complex state are harder to manage
+- Build/deployment includes both server-side build patterns and client-side build patterns
+- Hard to know where something is rendered
+- No clear conventions, no standard way to build them.
+- Can quickly turn into a ball of JavaScript with mixed frameworks
+- Generally uses two (or more) package managers
+
+### If your use case requires complex client-side interactivity, then you may need a single-page application.
```
Comment (ryanhofdotgov) [inline _guide/_pages/web-architecture.md:106]:
```suggestion
### If your use case requires complex client-side interactivity, then you may need a single-page application (SPA).
```

File: _guide/_pages/web-architecture.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+- Proper SEO also requires more developer time and effort
+- Deployments: how do you monitor when a new version of a SPA is available and apply the code changes? What if the SPA is a different version than the server?
+- An SPA can run for days, weeks, etc; which may highlight memory management problems
+- Routing: the browser already handles this, but SPA's override it and it becomes your problem
+- Caching can be tricky with many areas to maintain state storage (rather than just the browser’s cache)
+- Execution environment is always unknown and changing
+
+### Conclusion
+
+Keeping UI architecture as simple as possible can help keep government websites cost-effective, maintainable, and accessible. Understanding your user requirements can help decide what kinds of client-side interactivity are nice-to-haves, must-haves, or not needed at all.
```
Comment (ryanhofdotgov) [inline _guide/_pages/web-architecture.md:131]:
```suggestion
Keeping web application architecture as simple as possible can help keep government websites cost-effective, maintainable, and accessible. Understanding your user requirements can help decide what kinds of client-side interactivity are nice-to-haves, must-haves, or not needed at all.
```

**Review by ryanhofdotgov — APPROVED** [review]
I suggested some hyperlinks to make it more accessible to partner agency tech lead folks and flagged some verbiage that was tied to the old article title, but they don't prevent this from being article from being merged when you see fit.

For individual technologies, you could also specifically tag them with `DEFAULT` or `SUGGESTION` (see [Project Setup](https://engineering.18f.gov/project-setup/) for examples). 

Fantastic work – thank you all for this major contribution!

**Comment by echappen** [comment]
@ryanhofdotgov Thanks for all those links. Friday Brain didn't think of that.

**Review by alexsoble — APPROVED** [review]

**Comment by alexsoble** [comment]
Looks great! 👍

**Review by KendrickDaniel — APPROVED** [review]

**Pull request merged**